### PR TITLE
`SequenceIR` - convert `.flatten` into a pass

### PR DESCRIPTION
**Pull request by TsafrirA**
### Summary
#11951 introduced `SequenceIR` with a built it `.flatten()` method. It was later decided that all operations altering the IR should be performed as part of compiler passes. To this end, this PR converts the flatten method into a dedicated pass.

It should be noted that with `.flatten()` removed, `.draw()` can no longer draw the flattened version of the IR, as it could before (cyclic imports).

### CODE REVIEW

File: qiskit/pulse/compiler/passes/flatten.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            return None
+
+        if any(prog.time_table[x] is None for x in prog.sequence.node_indices() if x not in (0, 1)):
+            raise PulseCompilerError(
+                "Can not flatten unscheduled IR. Use SetSchedule pass before Flatten."
+            )
+
+        for ind in prog.sequence.node_indices():
+            if isinstance(sub_block := prog.sequence.get_node_data(ind), SequenceIR):
+                sub_block.flatten(inplace=True)
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/flatten.py]: Does this exist?

File: qiskit/pulse/compiler/passes/flatten.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                initial_time = prog.time_table[ind]
+                nodes_mapping = prog.sequence.substitute_node_with_subgraph(
+                    ind, sub_block.sequence, lambda x, y, _: edge_map(x, y, ind)
+                )
+                if initial_time is not None:
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/flatten.py]: Do you need this check?

**Review by nkanazawa1989 — CHANGES_REQUESTED** [review]
Very quick review. I'll review this again after this passes test.

File: qiskit/pulse/compiler/passes/flatten.py (hunk 1 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/flatten.py]: -facepalm-

File: qiskit/pulse/compiler/passes/flatten.py (hunk 2 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/flatten.py]: Good catch. Thanks!

File: qiskit/pulse/compiler/passes/flatten.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    def _flatten(self, prog: SequenceIR) -> SequenceIR:
+        """Recursively flatten the SequenceIR.
+
+        Returns:
+            A flattened ``SequenceIR`` object.
+
+        Raises:
+            PulseCompilerError: If ``prog`` is not scheduled.
+        """
+        # TODO : Verify that the block\sub blocks are sequenced correctly?
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/flatten.py]: Is this addressed by the sequence check in the _flatten function?

File: qiskit/pulse/compiler/passes/flatten.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        self,
+        passmanager_ir: SequenceIR,
+    ) -> SequenceIR:
+        """Run the pass."""
+
+        self._flatten(passmanager_ir)
+        return passmanager_ir
+
+    # pylint: disable=cell-var-from-loop
+    def _flatten(self, prog: SequenceIR) -> SequenceIR:
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/flatten.py]:
```suggestion
    def _flatten_recursion(self, prog: SequenceIR) -> SequenceIR:
```
To align with your convention.

**Review by nkanazawa1989 — COMMENTED** [review]
Minor comments. This looks good to go.

File: qiskit/pulse/compiler/passes/flatten.py (hunk 3 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/flatten.py]: I wasn't sure if we should verify the validity of the sequences. Now that I think of it, we verify that everything is scheduled, and that should be enough.

**Review by nkanazawa1989 — APPROVED** [review]
LGTM

**Pull request merged**